routing-daemon/lib/openshift/routing/daemon.rb

```diff
@@ -333,65 +349,75 @@ def delete_application app_name, namespace
-      create_application app_name, namespace if @lb_controller.pools[pool_name].nil?
-      unless (types & @endpoint_types).empty?
-        @logger.info "Adding new member #{gear_host}:#{gear_port} to pool #{pool_name}"
-        @lb_controller.pools[pool_name].add_member gear_host, gear_port.to_i
-      else
-        @logger.info "Ignoring endpoint with types #{types.join(',')}"
+      with_pool app_name, namespace do |pool|
+        unless (types & @endpoint_types).empty?
+          @logger.info "Adding new member #{gear_host}:#{gear_port} to pool #{pool.name}"
+          pool.add_member gear_host, gear_port.to_i
+        else
+          @logger.info "Ignoring endpoint with types #{types.join(',')}"
+        end
       end
     end
 
     def remove_endpoint app_name, namespace, gear_host, gear_port
-      pool_name = generate_pool_name app_name, namespace
-      member = gear_host + ':' + gear_port.to_s
-      # The remove_endpoint notification does not include a "types" field, so we
-      # cannot simply filter out endpoint types that we don't care about.
-      # Furthermore, the daemon by design does not store state itself, so we
-      # cannot check keep track of endpoints that we did not add to the load
-      # balancer.  All we can do is check whether the endpoint exists and delete
-      # it if it does.
-      if @lb_controller.pools[pool_name].members.include?(member)
-        @logger.info "Deleting member #{member} from pool #{pool_name}"
-        @lb_controller.pools[pool_name].delete_member gear_host, gear_port.to_i
-      else
-        @logger.info "No member #{member} exists in pool #{pool_name}; ignoring"
+      with_pool app_name, namespace, false do |pool|
+        member = gear_host + ':' + gear_port.to_s
+        if not pool.members.include?(member)
+          @logger.info "Pool member #{member} not found in pool #{pool.name}; resynching pool..."
+          pool.synch
+        end
+
+        # The remove_endpoint notification does not include a "types" field, so we
+        # cannot simply filter out endpoint types that we don't care about.
+        # Furthermore, the daemon by design does not store state itself, so we
+        # cannot check keep track of endpoints that we did not add to the load
+        # balancer.  All we can do is check whether the endpoint exists and delete
+        # it if it does.
+        if pool.members.include?(member)
+          @logger.info "Deleting member #{member} from pool #{pool.name}"
+          pool.delete_member gear_host, gear_port.to_i
+        else
+          @logger.info "No member #{member} exists in pool #{pool.name}; ignoring"
+        end
       end
     end
 
     def add_alias app_name, namespace, alias_str
-      pool_name = generate_pool_name app_name, namespace
-      @logger.info "Adding new alias #{alias_str} to pool #{pool_name}"
-      @lb_controller.pools[pool_name].add_alias alias_str
+      with_pool app_name, namespace do |pool|
```

Comment:
It makes sense to me that all the `add_` methods should as a rule create the pool if it does not exist.  Using the default of `create_if_nil=true` here does indeed change the behaviour, but I see no reason for `add_alias` to be an exception to that rule.  There is a risk if events are handled out of order (for example, by clustered routing-daemon instances, each picking event notifications off a queue), and in taking this approach, the result of handling operations out of order could be that we create an extra pool (for example, if we get a create-application event, an add-alias event, and a delete-application event and handle the last two out of order).  On the other hand, if we specified `create_if_nil=false`, then if corresponding create-application and add-alias events were swapped, the result would be that we would fail to create the pool and alias when we should have done so (the pool would be created by any subsequent add-public-endpoint event, but the alias would never be created).